pybricks/common/pb_type_speaker.c

```diff
@@ -22,34 +23,90 @@
 #include <pybricks/util_mp/pb_obj_helper.h>
 #include <pybricks/util_pb/pb_error.h>
 
+#define MP_PI MICROPY_FLOAT_CONST(3.14159265358979323846)
+
 typedef struct {
     mp_obj_base_t base;
     bool intialized;
 } pb_type_Speaker_obj_t;
 
 STATIC pb_type_Speaker_obj_t pb_type_Speaker_singleton;
 
+STATIC uint8_t volume = 100; // volume in 0..100 range
+
+// The number to multiply the sample amplitude by, to amplify the signal by the defined volume.
+// The original sample amplitude must be in the -1..1 range.
+STATIC uint16_t sample_amplifier = INT16_MAX;
+
 STATIC uint16_t waveform_data[128];
 
+// The amount of radians each PCM sample represents to completely fill whe waveform buffer for 2*PI
+//STATIC mp_float_t pcm_sample_size_rad = MP_PI * 2 / MP_ARRAY_SIZE(waveform_data);
+
+STATIC mp_obj_t pb_type_Speaker_get_volume(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    return mp_obj_new_int(volume);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(pb_type_Speaker_get_volume_obj, 1, pb_type_Speaker_get_volume);
+
+STATIC mp_obj_t pb_type_Speaker_set_volume(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    PB_PARSE_ARGS_METHOD(n_args, pos_args, kw_args,
+        pb_type_Speaker_obj_t, self,
+        PB_ARG_DEFAULT_INT(volume, 100));
+
+    (void)self; // unused
+    mp_int_t new_volume = pb_obj_get_int(volume_in);
+
+    if (volume <= 100) {
+        volume = new_volume;
+    } else {
+        volume = 100;
+    }
+
+    // linear amplification
+//    sample_amplifier = volume * INT16_MAX / 100;
+
+    // exponential amplification (human ear perceives sample amplitude in a logarithmic way)
+    sample_amplifier = (powf(10, volume / 100.0) - 1) / 9 * INT16_MAX;
+
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(pb_type_Speaker_set_volume_obj, 1, pb_type_Speaker_set_volume);
+
+STATIC void pb_type_Speaker_generate_square_wave(void) {
+    uint16_t lo_amplitude_value = INT16_MAX - sample_amplifier;
+    uint16_t hi_amplitude_value = sample_amplifier + INT16_MAX;
+
+    int i = 0;
+    for (; i < MP_ARRAY_SIZE(waveform_data) / 2; i++) {
+        waveform_data[i] = lo_amplitude_value;
+    }
+    for (; i < MP_ARRAY_SIZE(waveform_data); i++) {
+        waveform_data[i] = hi_amplitude_value;
+    }
+}
+
+//STATIC void pb_type_Speaker_generate_sine_wave(void) {
+//    for (int i = 0; i < MP_ARRAY_SIZE(waveform_data); i++) {
+//        waveform_data[i] = sinf(i * pcm_sample_size_rad) * sample_amplifier + INT16_MAX;
+//    }
+//}
+
+// For 0 frequencies that are just flat lines.
+STATIC void pb_type_Speaker_generate_line_wave(void) {
```

Comment:
If the frequency is 0, we should probably just turn off the DAC and audio amplifier instead of spending energy playing a silent sound. Not sure why I did it this way in the first place.

Reply:
Yes, will re-factor it.

Reply:
The flat line is used when playing `R` note. Is there a way to do that `pause` differently?

Reply:
Ah, that is probably why I did it that way. I think we can leave it for now and open an issue to remind us to look at this again later.